Add support to use wine to run cross compiled tests for cmake

rhabacker opened the pull request:
blocks #72

pali commented on CMakeLists.txt:
```diff
@@ -6,8 +6,19 @@ endif ()
+
 option(BUILD_SHARED_LIBS "shared/static libs" ON) 
 option(BUILD_TESTS "tests?" OFF)
+if(CMAKE_CROSSCOMPILING)
+    add_auto_option(ENABLE_WINE "Enable running tests with wine" AUTO)
+    find_program(WINE_EXECUTABLE wine)
+    check_auto_option(ENABLE_WINE "wine support" WINE_EXECUTABLE "wine executable")
```
Logic for usage wine should be: we are not running cmake on windows system (including cygwin) && we are running cmake on unix system && we are cross compiling for windows

pali commented on tests/CMakeLists.txt:
```diff
@@ -8,4 +8,4 @@ add_library(testdll3 SHARED testdll3.c)
 add_executable(t_dlfcn test.c)
 target_link_libraries(t_dlfcn dl)
-add_test(NAME t_dlfcn COMMAND $<TARGET_FILE:t_dlfcn> WORKING_DIRECTORY $<TARGET_FILE_DIR:t_dlfcn>)
+add_test(NAME t_dlfcn COMMAND ${WRAPPER} $<TARGET_FILE:t_dlfcn> WORKING_DIRECTORY $<TARGET_FILE_DIR:t_dlfcn>)
```
If we are cross compiling and wine is not used then we should not run test. But we should compile test (compilation is also some kind of test that it can be compiled).

rhabacker commented:
I think that the patches for this pr are now complete. If there are still any issues let me know. If not, please merge, so that this support can be used and extended by https://github.com/dlfcn-win32/dlfcn-win32/pull/72

traversaro commented on CMakeLists.txt:
```diff
@@ -6,11 +6,35 @@ endif ()
 
 project (dlfcn-win32 C)
 
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake/modules")
```
Using `CMAKE_SOURCE_DIR` is not indicate because it would not work if the project is included via `add_subdirectory`, see https://github.com/dlfcn-win32/dlfcn-win32/issues/49 . I suggest to use either `CMAKE_CURRENT_SOURCE_DIR` or `PROJECT_SOURCE_DIR` .

traversaro reviewed (changes requested):
A small request.

pali commented:
Why is travis configuration split into two files? Travis itself is running supplied script in bash, so there is no need to create extra bash script file.

pali commented:
And I must admit one thing: It seems that this change is huge, it adds 150 new lines which would need to be maintained. And in current form is cross compilation hardcoded/implemented only two cross compiler names. So it does not allow to use e.g. `i586-mingw32msvc-gcc` cross compiler. Why there is not just simple --cross-prefix switch or respecting now de-facto standard env variable `CROSS_COMPILE=`?

rhabacker commented:
> Why is travis configuration split into two files? Travis itself is running supplied script in bash, so there is no need to create extra bash script file.

With the additional script you can run the test on a local system very easily e.g 
```
ci_target=i686-w64-mingw32 ci_buildsys=cmake tools/ci-build.sh 
```
or 
```
ci_target=x86_64-w64-mingw32 ci_buildsys=Makefile tools/ci-build.sh 
```
which is very useful for local debugging, where inside the .travis.yml you cannot run that easily the configured tests.
Also I found that the bash support inside the .travis.yml file is very limited.

pali commented:
And I see there one difference on travis between Makefile and cmake build system:
Makefile show gcc warning for i686-w64-mingw32-gcc, but cmake does not show any warning for i686-w64-mingw32-gcc build.

So it looks like that cmake does not set -W -Wall flags to gcc

rhabacker commented:
> And I must admit one thing: It seems that this change is huge, it adds 150 new lines which would need to be maintained.

This has been reduced to about 40 lines, and you get an advantage. The cmake build system is now tested in the same way as the Makefile-based system and uses (except for cmake) the same dependencies. Because of the different test environments it was (and still is) sometimes difficult to find the cause of problems.

pali commented:
Nothing wrong with it, just I'm saying that for people who maintain lot of projects can be additional maintenance cost sometime problematic. But if @traversaro is fine with it, I have no problem.

For me it looks like quite suspicious that is needed extra bash script for running cmake.

Anyway, cmake is printing warnings on travis:
```
CMake Warning:
  Manually-specified variables were not used by the project:
    CMAKE_FIND_ROOT_PATH
    CMAKE_FIND_ROOT_PATH_MODE_LIBRARY
```

rhabacker commented:
> CMake Warning:
>   Manually-specified variables were not used by the project:
>     CMAKE_FIND_ROOT_PATH
>     CMAKE_FIND_ROOT_PATH_MODE_LIBRARY
> ```
https://stackoverflow.com/questions/36451368/get-rid-of-cmake-warning-manually-specified-variables-were-not-used-by-the-proj

rhabacker commented:
> And I see there one difference on travis between Makefile and cmake build system:
> Makefile show gcc warning for i686-w64-mingw32-gcc, but cmake does not show any warning for i686-w64-mingw32-gcc build.
> 
> So it looks like that cmake does not set -W -Wall flags to gcc

This realization is a first result for the advantages of this extension :-)

rhabacker commented:
> For me it looks like quite suspicious that is needed extra bash script for running cmake.

I think more people have knowledge about normal bash scripting than about the limited bash support within the .travis.yml file, which requires that you have to know what works and what doesn't every time. At first I tried this, but failed several times and then I remembered how other projects built this up, e.g. https://gitlab.freedesktop.org/dbus/dbus/-/blob/master/.travis.yml.

pali commented:
Ok!

So @traversaro, it is now up to you.

traversaro commented:
Thanks @rhabacker ! I see a few commits with Fixup in the message, do you want to curate the commits before the merge or I can simple Squash the commits and merge?

rhabacker commented:
> Thanks @rhabacker ! I see a few commits with Fixup in the message, do you want to curate the commits before the merge or I can simple Squash the commits and merge?

I squashed the fixme's into the related commits, add 'cmake:' prefix to some commits to be clearer and fixed spelling in one commit message. 

Also please note that in the meantime I recognized that  there was one hardcoded sub dir used in ci-bulid.sh to specifiy install_prefix:

-install_prefix=$(${ci_target}-gcc --print-sysroot)/mingw
+install_prefix=$(${ci_target}-gcc --print-sysroot)/${ci_target}

which I fixed too. See https://github.com/dlfcn-win32/dlfcn-win32/pull/85/commits/90302fc5e74bed533f68dbedb25163e98906121d

rhabacker commented:
Hi @traversaro, is there anything open to do ?

traversaro commented:
> 
> 
> Hi @traversaro, is there anything open to do ?

No, sorry, I just forgot to merge, sorry for the delay. Thanks a lot for the contribution!
